src/v/cluster/types.h

```diff
@@ -1343,4 +1371,28 @@ struct adl<cluster::commit_group_tx_reply> {
         return cluster::commit_group_tx_reply{ec};
     }
 };
+
+template<>
+struct adl<cluster::prepare_group_tx_request> {
+    void to(iobuf& out, cluster::prepare_group_tx_request&& r) {
+        serialize(
+          out,
+          std::move(r.ntp),
+          std::move(r.group_id),
+          r.etag,
+          r.pid,
+          r.tx_seq,
+          r.timeout);
+    }
+    cluster::prepare_group_tx_request from(iobuf_parser& in) {
+        auto ntp = adl<model::ntp>{}.from(in);
```

Comment:
even assuming that -O3 won't optimize that away, i don't think we should write that. it's like writing foo(read(), read()) and that isn't safe, and also ties the the adl implementation to the order of the fields in the structure (exactly what adl specializations allow us to avoid).